File: packages/ui/src/domain/issuerDomainBinding.ts
/**
 * The issuer↔domain DNS TXT binding, as the portals consume it.
 *
 * Mirrors the Rust `dogtag-dns-rs` wire shape. See that crate for the normative record convention:
 *
 *   `<clone-address-lowercase>._dogtag.<domain>  IN TXT  "<free-form description>"`
 *
 * The address is in the NAME so the VALUE can stay free-form — the domain owner writes whatever
 * description they like, and presence of the record at that name IS the assertion.
 *
 * # What the badge may and may not claim
 *
 * A `verified` binding proves ONE thing: the owner of that DNS zone published a record naming this
 * contract address. It does NOT establish that the named organisation is who it says it is — that comes
 * from the separate KYC/whitelist root of trust. Copy must state the observation, never a verdict about
 * the organisation or the credential. Credential validity is proven on-chain and is independent of this.
 *
 * # The binding is a THREE-LINK CHAIN
 *
 * All three links are required, and each is re-checked at verification time rather than inherited from
 * a stored claim:
 *
 *  1. **Factory provenance** — the contract was deployed by the DogTag factory, so it passed through
 *     KYC-gated `createIssuer`. Without this the rest is worthless: anyone can deploy their own
 *     contract, claim a domain for it and publish a matching TXT record. DNS would agree, the registry
 *     would agree, and none of it would mean anything.
 *  2. **The on-chain domain claim** — read from the domain registry, never from the document.
 *  3. **The DNS half** — that domain's zone lists this contract address.
 *
 * # Why six states and not three
 *
 * The three DNS outcomes (`verified` / `notListed` / `couldNotCheck`) must never collapse into each
 * other — conflating a resolver failure with an absence is the fail-open bug this feature exists to
 * remove. Three more exist because they answer genuinely different questions:
 *
 *  - `notADogTagIssuer` — link 1 FAILED. A categorically stronger statement than a missing DNS record,
 *    and it must never be softened into "not listed".
 *  - `noDomainClaimed` — link 1 holds and this issuer has claimed no domain. Nothing to check. The
 *    NORMAL state on day one, and it must look unremarkable.
 *  - `noDomainListed` — a DIRECTORY listing published no domain. No chain read happened at all, so it
 *    must never borrow `noDomainClaimed`'s on-chain wording.
 *  - `unavailable` — a prerequisite could not be read at all (nothing configured, or the read failed).
 *
 * Visually the three "we do not know" states share a neutral treatment with DIFFERENT text. That is not
 * the forbidden collapse: what is forbidden is letting `couldNotCheck` read as `verified` or as
 * `notListed`, or letting a provenance failure read as a DNS one.
 */

/** The state machine, exactly as the API emits it. `pending` is client-only (see below). */
export type IssuerDomainBindingState =
  | "verified"
  /** Link 1 failed: the contract was not deployed by the DogTag factory. */
  | "notADogTagIssuer"
  | "notListed"
  | "couldNotCheck"
  | "noDomainClaimed"
  /**
   * Directory-only: a provider-directory listing carried no domain. Never sent by the API.
   *
   * A directory row is not a chain read, so an absent domain column is evidence of nothing about what
   * the issuer published on-chain. `noDomainClaimed` states that the on-chain claim WAS read and was
   * empty; using it here would assert a read that never happened.
   */
  | "noDomainListed"
  | "unavailable"
  /**
   * Client-only: the lookup is in flight. Never sent by the API.
   *
   * A surface that renders before the chain read and DNS resolution complete MUST show this rather than
   * pre-filling a guess — showing `noDomainClaimed` first and then flipping would be synthesizing a
   * state that was never observed.
   */
  | "pending";

export interface IssuerDomainBinding {
  state: IssuerDomainBindingState;
  /** The ON-CHAIN claimed domain that was queried. Never the document's `issuer.domain`. */
  domain?: string;
  /** The clone address whose binding was sought (lowercase). */
  cloneAddress?: string;
  /** The fully-qualified TXT name queried, so an operator can reproduce the lookup by hand. */
  txtName?: string;
  /** Free-form text the DOMAIN OWNER published. Present only when `state === "verified"`. */
  description?: string;
  /** Unix seconds of the real DNS observation. A cached result keeps its original timestamp. */
  checkedAt?: number;
  /** Why the check could not run (`unavailable`), for an operator-facing tooltip. */
  detail?: string;

  // ---- block anchoring ----
  //
  // DNS changes and clones get superseded, so a binding without a block anchor is not auditable.
  /** The block every ON-CHAIN read here was pinned to. `null`/absent == the head could not be read. */
  blockNumber?: number | null;
  /** The block the issuer's domain CLAIM was written at — the anchor for "what did it claim at block N". */
  claimUpdatedAtBlock?: number;
  /** Unix seconds the claim was written at. */
  claimUpdatedAt?: number;
  /** Which authorized key wrote the claim. */
  claimSetBy?: string;
  /**
   * THE ASYMMETRY. `"live"` means this DNS answer was observed just now; `"stored"` means it is a
   * recorded past observation being replayed.
   *
   * Chain state is reproducible at any block with an archive node. DNS has NO history — there is no way
   * to ask what a zone published at block N, so a DNS answer can only ever be recorded, never
   * recomputed. A stored observation must never be shown as live, and a live one must never be shown as
   * proving the past.
   */
  dnsObservation?: "live" | "stored";
  /** Always false: no DNS answer can be historical. Present so the claim is explicit, not implied. */
  dnsHistorical?: boolean;
}

/** True only for a `verified` binding. Use this rather than hand-rolling a truthiness check. */
export function isDomainVerified(b: IssuerDomainBinding | null | undefined): boolean {
  return b?.state === "verified";
}

/** The visual register a state gets. Only `verified` is positive; only `notListed` is negative. */
export type BindingTone = "positive" | "negative" | "neutral" | "pending";

export function bindingTone(state: IssuerDomainBindingState): BindingTone {
  switch (state) {
    case "verified":
      return "positive";
    case "notListed":
      // Red plus a factual sentence, never an alarm. A missing record means the domain owner has not
      // published the binding; it says nothing about the credential, whose validity is separately
      // proven on chain.
      return "negative";
    case "notADogTagIssuer":
      // Also red, but its COPY is categorically different — see `bindingLine`. This is a definitive
      // observation about the contract's provenance, not about a DNS zone.
      return "negative";
    case "pending":
      return "pending";
    default:
      // `couldNotCheck`, `noDomainClaimed`, `noDomainListed` and `unavailable` are all "we do not
      // know". Deliberately NEITHER green nor red: a resolver timeout is not evidence either way.
      return "neutral";
  }
}

/**
 * The one-line copy for a state. Every string states what was LOOKED AT and what was FOUND, and stops
 * there — no "VERIFICATION FAILED", no "INVALID", no "UNTRUSTED", nothing that reads as a judgement on
 * the credential or the organisation.
 */
export function bindingLine(b: IssuerDomainBinding): string {
  const domain = b.domain?.trim() || "this domain";
  switch (b.state) {
    case "verified":
      return `This address is listed in ${domain}'s DNS records`;
    case "notListed":
      return `This address is not listed in ${domain}'s DNS records`;
    case "notADogTagIssuer":
      // Deliberately says nothing about DNS: DNS was never the question here.
      return "This contract was not deployed by the DogTag factory";
    case "couldNotCheck":
      return "We could not reach DNS to check this domain";
    case "noDomainClaimed":
      return "This issuer has published no domain on-chain";
    case "noDomainListed":
      return "No domain listed for this provider";
    case "unavailable":
      return "The on-chain domain claim could not be read";
    case "pending":
      return "Checking this domain's DNS records…";
  }
}

/**
 * The longer explanation, for a tooltip or a detail row. Says exactly what the check establishes and
 * what it deliberately does not.
 */
export function bindingExplanation(b: IssuerDomainBinding): string {
  switch (b.state) {
    case "verified":
      return (
        `The owner of ${b.domain} published a DNS TXT record naming this contract address. ` +
        "That shows the domain owner vouches for this address — it is not a check of the " +
        "organisation's identity, which comes from KYC at onboarding."
      );
    case "notListed":
      return (
        `${b.domain} publishes no DNS record for this contract address. Most issuers have not ` +
        "published one. This says nothing about the credential, whose validity is proven on-chain."
      );
    case "notADogTagIssuer":
      return (
        "The contract that issued this credential does not come from the DogTag factory, so it never " +
        "went through onboarding. A domain binding on such a contract would show only that whoever " +
        "deployed it also controls a domain."
      );
    case "couldNotCheck":
      return (
        "The DNS lookup did not complete, so we do not know whether the record exists. " +
        "This is evidence of nothing either way."
      );
    case "noDomainClaimed":
      return (
        "This issuer's contract has not claimed a domain on-chain, so there is nothing to check " +
        "against DNS. This is normal."
      );
    case "noDomainListed":
      return (
        "This provider's directory listing carries no domain. Nothing on-chain was read here, so " +
        "this says nothing about what the issuer may have published on-chain."
      );
    case "unavailable":
      return b.detail?.trim()
        ? `The on-chain domain claim could not be read: ${b.detail}`
        : "The on-chain domain claim could not be read.";
    case "pending":
      return "Reading the on-chain domain claim and resolving its DNS records.";
  }
}

/**
 * The provenance suffix for a binding line: which block the chain half came from, and that the DNS half
 * was seen live and cannot be re-derived for any past block.
 *
 * Returns `null` when there is nothing honest to say (no anchor, or a state with no DNS half).
 */
export function bindingProvenanceLine(b: IssuerDomainBinding): string | null {
  const parts: string[] = [];
  if (typeof b.blockNumber === "number") parts.push(`chain read at block ${b.blockNumber}`);
  const hasDnsHalf = b.state === "verified" || b.state === "notListed" || b.state === "couldNotCheck";
  if (hasDnsHalf) {
    // Say plainly which kind of observation this is. DNS has no history, so "live" is not a boast —
    // it is the only thing a DNS answer can ever be.
    parts.push(
      b.dnsObservation === "stored"
        ? "DNS as recorded earlier (DNS has no history, so it cannot be re-checked for the past)"
        : "DNS checked just now (DNS has no history, so it cannot be re-checked for the past)",
    );
  }
  return parts.length ? parts.join(" · ") : null;
}

/** The exact TXT record an issuer must publish, for operator-facing "how do I fix this?" copy. */
export function expectedTxtName(cloneAddress: string, domain: string): string {
  return `${cloneAddress.trim().toLowerCase()}._dogtag.${domain.trim().replace(/\.$/, "").toLowerCase()}`;
}

// -------------------------------------------------------------------------------------------------
// The issuer identity that accompanies the badge
// -------------------------------------------------------------------------------------------------

/**
 * The issuer identity as `/v1/verify` reports it.
 *
 * `onchainName` is AUTHORITATIVE: it is `DogTagIssuer.name()`, written by the factory's `onlyOwner`
 * `createIssuer` at KYC time. `documentName`/`documentDomain` come from the credential's `issuer` block,
 * which is OUTSIDE the Merkle root — a genuine credential can be re-badged as any authority and still
 * pass integrity. Never render the document values as the issuer; show them only to state a
 * disagreement.
 */
export interface IssuerIdentity {
  onchainName?: string | null;
  onchainNameAvailable?: boolean;
  /**
   * Link 1: "factoryDeployed" | "notFactoryDeployed" | "unknown".
   *
   * WHY the on-chain name is or is not available, and the two withheld cases are NOT the same fact:
   * `notFactoryDeployed` means the contract WAS read and the chain says it did not come from the
   * factory, while `unknown` means nothing was read at all. Reporting either as the other states
   * something that did not happen.
   */
  provenance?: string;
  documentName?: string;
  documentDomain?: string;
  rootCoveredDomain?: string | null;
  /** "match" | "mismatch" | "notAssertable" — never a boolean. */
  assertion?: string;
  documentNameDiffers?: boolean;
  documentDomainDiffers?: boolean;
  relabelled?: boolean;
}

/**
 * The issuer name a surface should display, whether it is the authoritative one, and the short label
 * that says where it came from.
 *
 * The label is returned FROM HERE rather than written at each call site so a surface cannot render a
 * name without saying where it came from, and so two surfaces cannot describe the same state
 * differently. Each label is an OBSERVATION about the read, never a verdict about the organisation:
 * a name that is not authoritative says nothing about the credential, whose validity is separately
 * proven on-chain.
 */
export function displayIssuerName(id: IssuerIdentity | null | undefined): {
  name: string;
  authoritative: boolean;
  sourceLabel: string;
} {
  const onchain = id?.onchainName?.trim();
  if (id?.onchainNameAvailable && onchain) {
    return { name: onchain, authoritative: true, sourceLabel: "(from the issuing contract)" };
  }
  return {
    name: id?.documentName?.trim() || "Unknown issuer",
    authoritative: false,
    sourceLabel:
      id?.provenance === "notFactoryDeployed"
        ? "(from the document — the issuing contract was not deployed by the DogTag factory, so its own name is not authoritative)"
        : "(from the document — the issuing contract could not be read)",
  };
}

File: apps/ios/DogTag/IssuerDomainBinding.swift
import Foundation

/// The issuer↔domain binding, resolved on the phone.
///
/// # The three-link chain
///
/// All three links are required, and all three are checked HERE at display time rather than inherited
/// from anything a server or a document says:
///
///  1. **Factory provenance** — `DogTagIssuerFactory.isClone(clone)`. Without it the rest is worthless:
///     anyone can deploy their own contract, claim a domain for it and publish a matching TXT record.
///     DNS would agree, the registry would agree, and none of it would mean anything, because that
///     contract never passed through the KYC-gated `createIssuer`.
///  2. **The on-chain domain claim** — `IssuerDomainRegistry.domainOf(clone)`. Read from the CHAIN, never
///     from the credential: the document's `issuer` block is outside the Merkle root, so a relabelled
///     document verifies. Taking the domain from the chain removes that attack entirely, because a
///     relabelled document cannot move a value it does not carry.
///  3. **The DNS half** — that domain's zone publishes
///     `<clone-address-lowercase>._dogtag.<domain> TXT "<free-form description>"`.
///
/// # Where the DNS lookup happens, and what it costs
///
/// The phone resolves DNS-over-HTTPS DIRECTLY, with no DogTag server in the loop. The alternative —
/// routing the lookup through our own API — would tell US which credential the holder is looking at and
/// when, which is strictly worse for a privacy-sensitive wallet than telling a public resolver a domain
/// name. This also matches the existing precedent: the phone already resolves DoH directly for the
/// pre-disclosure groomer binding (`ScanScreen`).
///
/// The privacy cost, stated plainly: **Cloudflare learns the issuer's domain and this device's IP at the
/// moment a credential is viewed.** It does not learn the credential, the holder, or the dogTagId.
///
/// # No synthesized states
///
/// Every state below is the outcome of a real read or a real read that really failed. Nothing is
/// defaulted, and a surface that renders before resolution completes shows `.pending` rather than
/// guessing — flashing "no domain claimed" and then flipping would be showing a state never observed.
enum IssuerBindingState: Equatable {
    /// The domain publishes a record for this contract. `description` is the domain owner's own words.
    case verified(description: String)
    /// Link 1 failed: this contract did not come from the DogTag factory. Categorically stronger than a
    /// missing DNS record and must never be shown as one.
    case notADogTagIssuer
    /// Links 1 and 2 hold; the domain publishes no record for this address. A NORMAL state.
    case notListed
    /// A real DNS attempt failed. Evidence of nothing.
    case couldNotCheck
    /// Link 1 holds; this issuer has claimed no domain on-chain. The normal day-one state.
    case noDomainClaimed
    /// A prerequisite could not be read at all (nothing configured, or the read failed).
    case unavailable
    /// In flight. Never a resting state.
    case pending
}

/// A resolved binding plus the provenance needed to be honest about it.
struct IssuerBinding: Equatable {
    var state: IssuerBindingState = .pending
    /// The ON-CHAIN claimed domain that was queried.
    var domain: String = ""
    /// The clone's on-chain `name()` — the only authoritative issuer name.
    var onchainName: String = ""
    /// The block every on-chain read was pinned to. `nil` == the head could not be read, so this answer
    /// is not reproducible and must not imply a block.
    var blockNumber: UInt64?
    /// When the DNS half was observed. DNS has no history, so this is the only timestamp it can carry.
    var checkedAt: Date?

    static let pending = IssuerBinding()

    var isVerified: Bool {
        if case .verified = state { return true }
        return false
    }
}

/// Resolves the binding for a credential's issuer contract.
enum IssuerBindingResolver {
    /// The underscore label that namespaces every DogTag binding record. An underscore label cannot
    /// collide with a real hostname, so the binding can never shadow something the domain owner serves.
    /// Address-label-first mirrors DKIM's `<selector>._domainkey.<domain>`.
    static let label = "_dogtag"

    /// A small in-process cache. A DNS lookup per row render is unacceptable, and a cached entry is a
    /// real prior observation — it keeps its ORIGINAL `checkedAt`, so a surface shows when DNS was really
    /// consulted rather than implying it just looked.
    private static var cache: [String: (binding: IssuerBinding, expires: Date)] = [:]
    private static let answerTtl: TimeInterval = 900   // 15 min for a real answer
    private static let failureTtl: TimeInterval = 30   // a blip must clear on its own
    private static let lock = NSLock()

    /// The normative TXT name binding `clone` to `domain`. `nil` when either input is unusable, so the
    /// caller reports "could not check" instead of firing a nonsense query.
    static func txtName(clone: String, domain: String) -> String? {
        let addr = clone.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let dom = domain.trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .trimmingCharacters(in: CharacterSet(charactersIn: "."))
        guard addr.count == 42, addr.hasPrefix("0x"),
              addr.dropFirst(2).allSatisfy({ $0.isHexDigit }) else { return nil }
        guard dom.contains("."), !dom.contains("/"), !dom.contains(":"),
              !dom.contains(where: { $0.isWhitespace }) else { return nil }
        return "\(addr).\(label).\(dom)"
    }

    /// Resolve the full chain. `clone` is the issuing contract; prefer the address the CHAIN resolved for
    /// the credential's root over the document's `issuer.documentStore` where available.
    static func resolve(
        rpcUrl: String,
        factory: String,
        domainRegistry: String,
        clone: String,
        useCache: Bool = true
    ) async -> IssuerBinding {
        let key = "\(clone.lowercased())|\(domainRegistry.lowercased())|\(factory.lowercased())"
        if useCache, let hit = cached(key) { return hit }

        var out = IssuerBinding()
        out.blockNumber = await RoaxRpc.blockNumber(rpcUrl: rpcUrl)

        // ---- link 1: factory provenance --------------------------------------------------------
        guard !factory.isEmpty else {
            out.state = .unavailable
            store(key, out)
            return out
        }
        switch await RoaxRpc.isClone(rpcUrl: rpcUrl, factory: factory, candidate: clone, atBlock: out.blockNumber) {
        case .invalid:
            out.state = .notADogTagIssuer
            store(key, out)
            return out
        case .unknown:
            // The read did not resolve. NOT "not a DogTag issuer" — we do not know.
            out.state = .unavailable
            store(key, out)
            return out
        case .valid:
            break
        }

        // The authoritative issuer name, read once provenance holds.
        if case .value(let n) = await RoaxRpc.issuerOnchainName(rpcUrl: rpcUrl, clone: clone, atBlock: out.blockNumber) {
            out.onchainName = n
        }

        // ---- link 2: the on-chain domain claim -------------------------------------------------
        guard !domainRegistry.isEmpty else {
            out.state = .unavailable
            store(key, out)
            return out
        }
        switch await RoaxRpc.issuerClaimedDomain(
            rpcUrl: rpcUrl, domainRegistry: domainRegistry, clone: clone, atBlock: out.blockNumber
        ) {
        case .failure:
            out.state = .unavailable
            store(key, out)
            return out
        case .noContract:
            // An EMPTY eth_call result means no contract at that address — the registry is not deployed
            // for this config. That is "we could not check", never "this issuer claims no domain".
            out.state = .unavailable
            store(key, out)
            return out
        case .value(let d):
            let domain = d.trimmingCharacters(in: .whitespacesAndNewlines)
            if domain.isEmpty {
                // A real, ABI-encoded empty string: the issuer genuinely claims no domain.
                out.state = .noDomainClaimed
                store(key, out)
                return out
            }
            out.domain = domain
        }

        // ---- link 3: the DNS half ---------------------------------------------------------------
        out.checkedAt = Date()
        guard let name = txtName(clone: clone, domain: out.domain) else {
            out.state = .couldNotCheck
            store(key, out)
            return out
        }
        out.state = await resolveDns(name: name)
        store(key, out)
        return out
    }

    /// Resolve the TXT record over DoH. The classification itself is [`classifyDoh`], kept pure so the
    /// three-state rule is unit-testable without a network.
    private static func resolveDns(name: String) async -> IssuerBindingState {
        guard let encoded = name.addingPercentEncoding(
            withAllowedCharacters: CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-._~"))
        ) else {
            return .couldNotCheck
        }
        let url = "https://cloudflare-dns.com/dns-query?name=\(encoded)&type=TXT"
        let resp = await Http.getJSON(url, accept: "application/dns-json")
        guard resp.ok, let d = resp.body.data(using: .utf8),
              let o = (try? JSONSerialization.jsonObject(with: d)) as? [String: Any] else {
            return .couldNotCheck
        }
        return classifyDoh(o)
    }

    /// Turn a DoH JSON answer into one of the three DNS states, branching on the `Status` RCODE so a
    /// resolver failure can never collapse into an absence. This is the mirror of `dogtag-dns-rs`'s
    /// `classify_txt` and must stay behaviourally identical to it.
    static func classifyDoh(_ o: [String: Any]) -> IssuerBindingState {
        // An ABSENT Status is not "no error" — it means this is not a DoH JSON answer, and reading it as
        // NOERROR would turn a misconfiguration into a confident "absent".
        guard let status = o["Status"] as? Int else { return .couldNotCheck }
        switch status {
        case 0: break                      // NOERROR — inspect the answers
        case 3: return .notListed          // NXDOMAIN — a DEFINITIVE absence, not a failure
        default: return .couldNotCheck     // SERVFAIL(2) / REFUSED(5) / … — a non-answer
        }
        guard let answers = o["Answer"] as? [[String: Any]] else {
            return .notListed              // NOERROR with no Answer section (NODATA)
        }
        for a in answers {
            // type 16 == TXT; the Answer array may also carry CNAMEs from an alias chain.
            guard (a["type"] as? Int) == 16, let raw = a["data"] as? String else { continue }
            return .verified(description: unquoteTxt(raw))
        }
        return .notListed
    }

    /// Join a DoH TXT value into the string the zone published. TXT RDATA is a sequence of
    /// character-strings each at most 255 octets, rendered as space-separated quoted chunks; RFC 1035
    /// says a consumer CONCATENATES them, so `"abc" "def"` is `abcdef`, not `abc def`. Getting this wrong
    /// would corrupt any description over 255 bytes.
    static func unquoteTxt(_ raw: String) -> String {
        let s = raw.trimmingCharacters(in: .whitespaces)
        guard s.contains("\"") else { return s }
        var out = ""
        var inQuotes = false
        var escaped = false
        for ch in s {
            if escaped { out.append(ch); escaped = false; continue }
            if ch == "\\", inQuotes { escaped = true; continue }
            if ch == "\"" { inQuotes.toggle(); continue }
            if inQuotes { out.append(ch) }
            // whitespace between chunks is dropped, per concatenation semantics
        }
        return out
    }

    private static func cached(_ key: String) -> IssuerBinding? {
        lock.lock(); defer { lock.unlock() }
        guard let hit = cache[key], hit.expires > Date() else { return nil }
        return hit.binding
    }

    private static func store(_ key: String, _ b: IssuerBinding) {
        let ttl: TimeInterval
        switch b.state {
        case .couldNotCheck, .unavailable: ttl = failureTtl
        default: ttl = answerTtl
        }
        lock.lock(); defer { lock.unlock() }
        cache[key] = (b, Date().addingTimeInterval(ttl))
    }

    /// Drop every cached observation (used on an explicit refresh).
    static func clearCache() {
        lock.lock(); defer { lock.unlock() }
        cache.removeAll()
    }
}

// -------------------------------------------------------------------------------------------------
// Copy — the observation, never a verdict
// -------------------------------------------------------------------------------------------------

extension IssuerBinding {
    /// One factual line: what was looked at, and what was found. Deliberately contains no
    /// "VERIFICATION FAILED", "INVALID", "UNTRUSTED" or warning language — a missing DNS record means the
    /// domain owner has not published the binding, and says nothing about the credential, whose validity
    /// is separately and genuinely proven on-chain. Telling someone their valid credential failed would
    /// be worse than showing nothing.
    var line: String {
        let dom = domain.isEmpty ? "this domain" : domain
        switch state {
        case .verified:        return "This address is listed in \(dom)'s DNS records"
        case .notListed:       return "This address is not listed in \(dom)'s DNS records"
        case .notADogTagIssuer: return "This contract was not deployed by the DogTag factory"
        case .couldNotCheck:   return "We could not reach DNS to check this domain"
        case .noDomainClaimed: return "This issuer has published no domain on-chain"
        case .unavailable:     return "The on-chain domain claim could not be read"
        case .pending:         return "Checking this domain's DNS records…"
        }
    }

    /// The domain owner's own description, shown only for a verified binding — the whole reason the TXT
    /// value is free-form.
    var publishedDescription: String? {
        if case .verified(let d) = state {
            let t = d.trimmingCharacters(in: .whitespacesAndNewlines)
            return t.isEmpty ? nil : t
        }
        return nil
    }

    /// Which register the line is shown in. Only a verified binding is positive; only a definitive
    /// absence or a provenance failure is negative. Everything we simply do not know stays NEUTRAL — a
    /// resolver timeout is not evidence either way, so colouring it as a failure would be a lie of
    /// emphasis.
    enum Tone { case positive, negative, neutral, pending }

    var tone: Tone {
        switch state {
        case .verified: return .positive
        case .notListed, .notADogTagIssuer: return .negative
        case .pending: return .pending
        case .couldNotCheck, .noDomainClaimed, .unavailable: return .neutral
        }
    }
}
